File: node/examples/executable.js
const {performance, PerformanceObserver} = require('perf_hooks');

const perfObserver = new PerformanceObserver((items) => {
    console.log('\nPerformance status:')
    items.getEntries().forEach((entry) => console.log(`- ${entry.name} ${entry.duration}ms`))
    console.log();
});

perfObserver.observe({entryTypes: ["measure"], buffered: true})

const game_scanner = require("../lib");

try {
    performance.mark('start');
    const games = game_scanner.amazon.executable();
    performance.mark('end');
    console.log(games);
    performance.measure("amazon", "start", "end");
} catch (error) {
    console.error(error);
}

try {
    performance.mark('start');
    const games = game_scanner.blizzard.executable();
    performance.mark('end');
    console.log(games);
    performance.measure("blizzard", "start", "end");
} catch (error) {
    console.error(error);
}

try {
    performance.mark('start');
    const games = game_scanner.epicgames.executable();
    performance.mark('end');
    console.log(games);
    performance.measure("epicgames", "start", "end");
} catch (error) {
    console.error(error);
}

try {
    performance.mark('start');
    const games = game_scanner.gog.executable();
    performance.mark('end');
    console.log(games);
    performance.measure("gog", "start", "end");
} catch (error) {
    console.error(error);
}

try {
    performance.mark('start');
    const games = game_scanner.origin.executable();
    performance.mark('end');
    console.log(games);
    performance.measure("origin", "start", "end");
} catch (error) {
    console.error(error);
}

try {
    performance.mark('start');
    const games = game_scanner.riotgames.executable();
    performance.mark('end');
    console.log(games);
    performance.measure("riotgames", "start", "end");
} catch (error) {
    console.error(error);
}

try {
    performance.mark('start');
    const games = game_scanner.steam.executable();
    performance.mark('end');
    console.log(games);
    performance.measure("steam", "start", "end");
} catch (error) {
    console.error(error);
}

try {
    performance.mark('start');
    const games = game_scanner.ubisoft.executable();
    performance.mark('end');
    console.log(games);
    performance.measure("ubisoft", "start", "end");
} catch (error) {
    console.error(error);
}

File: tests/executable.rs
use std::any::type_name;

#[cfg(target_os = "windows")]
mod amazon {
    use super::*;

    #[test]
    fn executable() {
        let executable = game_scanner::amazon::executable().unwrap_or_default();

        assert_eq!(LAUNCHER_EXECUTABLE_PATH, type_of(&executable));
    }
}

#[cfg(not(target_os = "linux"))]
mod blizzard {
    use super::*;

    #[test]
    fn executable() {
        let executable = game_scanner::blizzard::executable().unwrap_or_default();

        assert_eq!(LAUNCHER_EXECUTABLE_PATH, type_of(&executable));
    }
}

mod epicgames {
    use super::*;

    #[test]
    fn executable() {
        let executable = game_scanner::epicgames::executable().unwrap_or_default();

        assert_eq!(LAUNCHER_EXECUTABLE_PATH, type_of(&executable));
    }
}

mod gog {
    use super::*;

    #[test]
    fn executable() {
        let executable = game_scanner::gog::executable().unwrap_or_default();

        assert_eq!(LAUNCHER_EXECUTABLE_PATH, type_of(&executable));
    }
}

mod origin {
    use super::*;

    #[test]
    fn executable() {
        let executable = game_scanner::origin::executable().unwrap_or_default();

        assert_eq!(LAUNCHER_EXECUTABLE_PATH, type_of(&executable));
    }
}

mod riotgames {
    use super::*;

    #[test]
    fn executable() {
        let executable = game_scanner::riotgames::executable().unwrap_or_default();

        assert_eq!(LAUNCHER_EXECUTABLE_PATH, type_of(&executable));
    }
}

mod steam {
    use super::*;

    #[test]
    fn executable() {
        let executable = game_scanner::steam::executable().unwrap_or_default();

        assert_eq!(LAUNCHER_EXECUTABLE_PATH, type_of(&executable));
    }
}

#[cfg(target_os = "windows")]
mod ubisoft {
    use super::*;

    #[test]
    fn executable() {
        let executable = game_scanner::ubisoft::executable().unwrap_or_default();

        assert_eq!(LAUNCHER_EXECUTABLE_PATH, type_of(&executable));
    }
}

fn type_of<T>(_: T) -> &'static str {
    type_name::<T>()
}

const LAUNCHER_EXECUTABLE_PATH: &str = "&std::path::PathBuf";
